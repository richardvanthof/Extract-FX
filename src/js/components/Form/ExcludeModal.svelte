<style lang="scss">
.exclusions {
	background: var(--bg-dark);
	border: var(--border-styling);
	border-radius: var(--border-radius-inner);
	position: relative;
	margin-top: 1em;
    overflow: hidden;
}

.exclusions-controls {
  background: var(--bg-light);
  display: flex;
  justify-content: space-between;
  padding: 0.1em;
  min-width: .5em 2em;
  box-shadow: 0px .5px 5px 2px rgba(0,0,0,0.05);
}

.exclusions-container {
	min-height: 8rem;
    margin: .4em;
	/* padding-bottom: 2em; */
}

.exclusions-placeholder {
	margin: 0;
	transform: translateY(3em);
	text-align: center;
}

.exclusions {
    border-radius: var(--border-radius-outer)
}

#remove-all-exclusions-btn {
	background: none;
	border:none;
	font-size: 0.85em;
	margin: 0 0.5em;
    &:active {
        background: var(--bg-btn-active)
    }
}

.exclusion-toolbar-button {
    padding: .66em 1.5em .66em 1em;
    border: none;
    border-right: var(--border-styling);
    background: none;
    border-radius: 0;
    &:active {
        background: var(--bg-btn-active)
    }
}

button, summary {
    cursor: pointer
}
</style>

<script>
    import Exclusion from "./Exclusion.svelte";
    import { v4 as uuidv4 } from 'uuid';

    let {exclusions} = $props();

    let setExclusions = {
        add: (effect) => exclusions.push({
            effect: effect || null,
            id: uuidv4()
        }),
        update: (newEffect, targetId) => {
            const index = exclusions.findIndex(({id}) => id === targetId)
            if(index) {
                exclusions[index] = {
                    id: targetId,
                    effect: newEffect
                }
            } else {
                throw 'Update failed. Exclusion not found.'
            }
        },
        remove: (targetId) => {
            exclusions = exclusions.filter(({id}) => id != targetId)
        },
        removeAll: () => {
            exclusions = [];
        }
    }

    const {add, update, remove, removeAll} = setExclusions;

    const handleClick = (e, callback) => {
        e.preventDefault()
        return callback;
    }

</script>


<details open>
    <summary class="exclusion-header" id='exclusion-header'>Exclude effects { (exclusions.length > 0) ? `(${exclusions.length})` : ''}</summary>
    <div class="exclusions">
        <div class='exclusions-controls'>
                <button class='exclusion-toolbar-button exclusion-control' id='add-exclusion-btn' onclick={(e) => handleClick(e, add())}>+ Add Exclusion</button>
                <button class='exclusion-control' id='remove-all-exclusions-btn' onclick={(e) => handleClick(e, removeAll())}>Clear all</button>
            </div>
        <div class="exclusions-container">
            {#each exclusions as exclusion}
            <Exclusion {exclusion} {remove}/>
            {/each}
        </div>
    </div>
</details>
