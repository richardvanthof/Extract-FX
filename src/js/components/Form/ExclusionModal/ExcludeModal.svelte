<style lang="scss">

.wrapper {
    margin-bottom: 1em;
}

.exclusions {
	background: var(--bg-dark);
	border: var(--border-styling);
	border-radius: var(--border-radius-inner);
	position: relative;
	margin-top: 1em;
    overflow: hidden;
}

.exclusions-controls {
  background: var(--bg-light);
  display: flex;
  justify-content: space-between;
  padding: 0.1em;
  min-width: .5em 2em;
  box-shadow: 0px .5px 5px 2px rgba(0,0,0,0.05);
}

.exclusions-container {
	min-height: 8rem;
    margin: .4em;
	/* padding-bottom: 2em; */
}

.exclusions-placeholder {
	margin: 0;
	transform: translateY(2.5em);
	text-align: center;
}

.exclusions {
    border-radius: var(--border-radius-outer)
}

#remove-all-exclusions-btn {
	background: none;
	border:none;
	font-size: 0.85em;
	margin: 0 0.5em;
    &:active {
        background: var(--bg-btn-active)
    }
}

.exclusion-toolbar-button {
    padding: .66em 1.5em .66em 1em;
    border: none;
    border-right: var(--border-styling);
    background: none;
    border-radius: 0;
    &:active {
        background: var(--bg-btn-active)
    }
}

button, summary {
    cursor: pointer
}
</style>

<script lang="ts">
    import type {Exclusion as ExclusionType } from '../../../global-vars/shared';
    import Exclusion from "../Exclusion/Exclusion.svelte";
    import Button from "../../Button.svelte";
    import { v4 as uuidv4 } from 'uuid';
    import {handleClick} from '../../../helpers/helpers';
    let { exclusions, open } = $props();
    
    const excl = $derived($exclusions);

    // Add a new exclusion item
    const add = () => {
        exclusions.update(
            (current: ExclusionType[]) => [
                ...current,
                {
                    effect: null,  // You might want to set this dynamically
                    id: uuidv4()
                }
            ]
        );
    };

    // Update an exclusion item
    const update = (newEffect: string, id: string) => {
        exclusions.update( (currentItems: ExclusionType[]) => {
            return currentItems.map(item => item.id === id ? {...item, effect: newEffect} : item)
        })
    }

    const remove = (id: string) => {
        console.log(`remove ${id}`);
        exclusions.update((currentItems: ExclusionType[]) => {
            return currentItems.filter(({ id: exclusionId }) => exclusionId !== id);
        });
    };

    // Remove all exclusions
    const removeAll = () => exclusions.set([]);
</script>

<details class="wrapper" bind:open={$open}>
    <Button title="test" />
    <summary class="exclusion-header" id='exclusion-header'>
        Exclude effects { ($exclusions.length > 0) ? `(${$exclusions.length})` : '' }
    </summary>
    <div class="exclusions">
        <div class='exclusions-controls'>
            <button 
                class='exclusion-toolbar-button exclusion-control' 
                id='add-exclusion-btn' 
                title="add"
                onclick={(e:Event) => {
                    e.preventDefault();
                    add()
                }}>+ Add Exclusion
            </button>
            <button 
                class='exclusion-control' 
                title="clear"
                id='remove-all-exclusions-btn' 
                onclick={(e:Event) => {
                    e.preventDefault();
                    removeAll()
                }}>Clear all</button>
        </div>
        <div class="exclusions-container">
            {#if $exclusions.length > 0}
                {#each excl as {id}}
                    <Exclusion {id} {exclusions} {remove}/>
                {/each}
            {:else}
                <div class="exclusions-placeholder">
                    <p class='caption'>Press the +-button to add an exclusion.</p>
                </div>
            {/if}
            
        </div>
    </div>
</details>
