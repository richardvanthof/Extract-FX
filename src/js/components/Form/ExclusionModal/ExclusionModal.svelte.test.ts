import { describe, expect, it } from 'vitest';
import { render, screen } from '@testing-library/svelte';
import '@testing-library/jest-dom';
import userEvent from '@testing-library/user-event'

import ExcludeModal from './ExcludeModal.svelte';

describe("Exclusion modal", () => {
    const options = ['Effect A', 'Effect B', 'Effect C']
    const open = true;
    
    describe('Rendering', () => {
        it("Empty component loads", () => {
            const exclusions = $state([]);
    
            const {container} = render(ExcludeModal, {
                props: {exclusions, open},
                context: new Map([['exclusionOptions', options]])
            });
            
            const placeholder = screen.queryByText('Press the +-button to add an exclusion.');
            
            // Placeholder loads
            expect(placeholder).toBeInTheDocument()
            
            // Empty component loads
            expect(container).toMatchSnapshot()
            
        })
        
        it("Loads currect exclusion items state", () => {
            const exclusions = $state([
                {effect: 'Effect A',id: 'FX1'},
                {effect: 'Effect B', id: 'FX2'}
            ]);
            
            const {container} = render(ExcludeModal, {
                props: {exclusions, open},
                context: new Map([['exclusionOptions', options]])
            });
    
            exclusions.forEach(({effect}) => {
                const value = screen.getByDisplayValue(effect);
                expect(value).toBeInTheDocument()
            })        
            
            expect(container).toMatchSnapshot()   
        })
    })

    describe('Modifiy list', () => {
       
        const exclusions = $state([
            {effect: 'Effect A',id: 'FX1'},
            {effect: 'Effect B', id: 'FX2'}
        ]);

        const user = userEvent.setup();

        it("Add", async () => {

            render(ExcludeModal, {
                props: {exclusions, open},
                context: new Map([['exclusionOptions', options]])
            });

            const addBtn = screen.getByText('+ Add Exclusion')
            await user.click(addBtn);

            const newExclusion = screen.getByDisplayValue('Choose an effect');
            const allExclusions = screen.getAllByTestId('exclusion')
            expect(newExclusion).toBeInTheDocument();
            expect(allExclusions.length).toBe(3);
        })

        it("Delete", async () => {

            render(ExcludeModal, {
                props: {exclusions, open},
                context: new Map([['exclusionOptions', options]])
            });

            const deleteBtn = screen.getAllByTestId('delete-exclusion')
            await user.click(deleteBtn[0]);

            const allExclusions = screen.getAllByTestId('exclusion')
            expect(allExclusions.length).toBe(1); // Updated in DOM
        })

        it("Update", async () => {

            render(ExcludeModal, {
                props: {exclusions, open},
                context: new Map([['exclusionOptions', options]])
            });

            // Find all dropdowns (assuming they are <select> elements)
            const dropdowns = screen.getAllByTestId('select-exclusion');

            dropdowns[0].focus(); // Manually focus the dropdown if it's not focused yet
            user.click(dropdowns[0]);
            user.keyboard('ArrowDown');
            user.keyboard('Enter');

            // Check if the correct value has been selected
            expect(dropdowns[0]).toHaveValue('Effect B');
        })


        it("Delete All", async () => {
  
            render(ExcludeModal, {
                props: {exclusions, open},
                context: new Map([['exclusionOptions', options]])
            });

            const clearBtn = screen.getByText('Clear all')
            await user.click(clearBtn);

            // Check that no exclusions are rendered (i.e., the exclusion elements are removed)
            const allExclusions = screen.queryAllByTestId('exclusion');
            expect(allExclusions.length).toBe(0);  // Check that there are no exclusions
        })


    })

})