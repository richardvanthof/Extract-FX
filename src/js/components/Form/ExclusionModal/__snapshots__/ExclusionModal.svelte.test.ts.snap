// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Exclusion modal > Rendering > Empty component loads 1`] = `
<div>
  <details
    class="wrapper svelte-fh62sw"
    open=""
  >
    <summary
      class="exclusion-header svelte-fh62sw"
      id="exclusion-header"
    >
      Exclude effects 
    </summary>
     
    <div
      class="exclusions svelte-fh62sw"
    >
      <div
        class="exclusions-controls svelte-fh62sw"
      >
        <button
          class="exclusion-toolbar-button exclusion-control svelte-fh62sw"
          id="add-exclusion-btn"
          title="add"
        >
          + Add Exclusion
        </button>
         
        <button
          class="exclusion-control svelte-fh62sw"
          id="remove-all-exclusions-btn"
          title="clear"
        >
          Clear all
        </button>
      </div>
       
      <div
        class="exclusions-container svelte-fh62sw"
      >
        <div
          class="exclusions-placeholder svelte-fh62sw"
        >
          <p
            class="caption"
          >
            Press the +-button to add an exclusion.
          </p>
        </div>
        <!---->
      </div>
    </div>
  </details>
  
</div>
`;

exports[`Exclusion modal > Rendering > Loads currect exclusion items state 1`] = `
<div>
  <details
    class="wrapper svelte-fh62sw"
    open=""
  >
    <summary
      class="exclusion-header svelte-fh62sw"
      id="exclusion-header"
    >
      Exclude effects (2)
    </summary>
     
    <div
      class="exclusions svelte-fh62sw"
    >
      <div
        class="exclusions-controls svelte-fh62sw"
      >
        <button
          class="exclusion-toolbar-button exclusion-control svelte-fh62sw"
          id="add-exclusion-btn"
          title="add"
        >
          + Add Exclusion
        </button>
         
        <button
          class="exclusion-control svelte-fh62sw"
          id="remove-all-exclusions-btn"
          title="clear"
        >
          Clear all
        </button>
      </div>
       
      <div
        class="exclusions-container svelte-fh62sw"
      >
        <!---->
        <div
          class="exclusion svelte-1q5kcw3"
          data-testid="exclusion"
        >
          <!---->
          <select
            class="svelte-14mby5s"
            data-testid="select-exclusion"
            id="sourceTrack"
          >
            <option
              class="svelte-14mby5s"
              value=""
            >
              Choose an effect
            </option>
            <option
              class="svelte-14mby5s"
              value="Effect A"
            >
              Effect A
            </option>
            <option
              class="svelte-14mby5s"
              value="Effect B"
            >
              Effect B
            </option>
            <option
              class="svelte-14mby5s"
              value="Effect C"
            >
              Effect C
            </option>
            
          </select>
          
          <!---->
           
          <button
            class="svelte-vc9cas"
            data-testid="delete-exclusion"
            name="x"
            title="Delete exclusion"
          >
            x
          </button>
          <!---->
        </div>
        <div
          class="exclusion svelte-1q5kcw3"
          data-testid="exclusion"
        >
          <!---->
          <select
            class="svelte-14mby5s"
            data-testid="select-exclusion"
            id="sourceTrack"
          >
            <option
              class="svelte-14mby5s"
              value=""
            >
              Choose an effect
            </option>
            <option
              class="svelte-14mby5s"
              value="Effect A"
            >
              Effect A
            </option>
            <option
              class="svelte-14mby5s"
              value="Effect B"
            >
              Effect B
            </option>
            <option
              class="svelte-14mby5s"
              value="Effect C"
            >
              Effect C
            </option>
            
          </select>
          
          <!---->
           
          <button
            class="svelte-vc9cas"
            data-testid="delete-exclusion"
            name="x"
            title="Delete exclusion"
          >
            x
          </button>
          <!---->
        </div>
        
        <!---->
      </div>
    </div>
  </details>
  
</div>
`;
